client/src/components/dom/Auth/AuthModal/AuthModal.tsx

```diff
@@ -0,0 +1,246 @@
+    setFormErrors({ email: "", password: "", confirm: "" });
+  }
+
+  function switchActions() {
+    resetForm();
+    resetErrors();
+    props.setAuthAction(
+      props.authAction === AuthAction.Login ? AuthAction.Register : AuthAction.Login,
+    );
+  }
+
+  function closeModal(e: any) {
+    if (e.target === e.currentTarget) {
+      resetForm();
+      props.setAuthAction(AuthAction.Idle);
+    }
+  }
+
+  function verifyLoginInputs() {
+    let isValid = true;
+    resetErrors();
+
+    // Check if email is valid
+    if (!formDetails.email.includes("@")) {
+      isValid = false;
+      setFormErrors(prev => ({ ...prev, email: "Invalid email" }));
+    }
+    if (formDetails.password === "") {
+      isValid = false;
+      setFormErrors(prev => ({ ...prev, password: "Required" }));
+    }
+    return isValid;
+  }
+
+  function verifyRegisterInputs() {
+    let isValid = true;
+    resetErrors();
+
+    // Check password length
+    if (formDetails.password.length < 6) {
+      isValid = false;
+      setFormErrors(prev => ({ ...prev, password: "Password must be at least 6 characters" }));
+    }
+    // Check if passwords match
+    if (formDetails.password !== formDetails.confirm) {
+      isValid = false;
+      setFormErrors(prev => ({ ...prev, confirm: "Passwords do not match" }));
+    }
+    // Check if email is valid
+    if (!formDetails.email.includes("@")) {
+      isValid = false;
+      setFormErrors(prev => ({ ...prev, email: "Invalid email" }));
+    }
+    // Check if any fields are empty
+    for (const key in formDetails) {
+      if (formDetails[key] === "") {
+        isValid = false;
+        setFormErrors(prev => ({ ...prev, [key]: "Required" }));
+      }
+    }
+    return isValid;
+  }
+
+  async function handleLogout() {
+    const error = await supabase.auth.signOut();
+    props.setAuthAction(AuthAction.Idle);
+  }
+
+  async function handleLogin() {
+    // Check if inputs are valid
+    if (!verifyLoginInputs()) return;
+
+    // Send request to supabase
+    const { data, error } = await supabase.auth.signInWithPassword({
+      email: formDetails.email,
+      password: formDetails.password,
+    });
+    // Check for api errors
+    if (error) {
+      // console.log(error);
+
+      setFormErrors(prev => ({
+        ...prev,
+        email: "Invalid email or password",
+        password: "Invalid email or password",
+      }));
+      return;
+    }
+    resetForm();
+    resetErrors();
+    props.setAuthAction(AuthAction.Idle);
+  }
+
+  async function handleRegister() {
+    // Check if inputs are valid
+    if (!verifyRegisterInputs()) return;
+
+    // Send request to supabase
+    const { data, error } = await supabase.auth.signUp({
+      email: formDetails.email,
+      password: formDetails.password,
+    });
+    // Check for api errors
+    if (error) {
+      // console.log(error);
+      return;
+    }
+    // Check if account already exists
+    if (!data.user.identities.length) {
```

Comment:
this implies that the user was created w/o error in supabase, as we passed the error boundary defined on line 148.

If the registration was successful, then what is this notion of multiple identities? If we are not allowing this registration then we should also remove the extraneous row from the supabase DB assuming it was successfully added.

Reply:
I'm not sure why, and you can test this out yourself, but if you supabase.auth.signUp() with an existing email, it doesn't actually error, it returns a data object with an empty user.identity array. If the email is not already in the db, the user.identity array is populated with an entry.

Without this check, we can't confirm or deny to the user that the account already exists or not.

Reply:
Looking more into it, it appears to be intentional to prevent data leaks?

https://github.com/supabase/supabase-js/issues/296

Reply:
Great find! As per [this comment](https://github.com/supabase/supabase-js/issues/296#issuecomment-976200828) it seems like the best course of action will be to simply ignore the fact that the email is in use already -- and let them know an email was sent.

By default it will be an account recovery email if they're already signed up, or a proper confirmation email if not. Either way, we just direct them to their mail-box. It's not an us problem.